sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,455 @@
+# coding=utf-8
+
+from abc import ABC, abstractmethod
+
+from sympy.physics.mechanics.body import Body
+from sympy.physics.vector import Vector, dynamicsymbols, cross
+from sympy.physics.vector.frame import ReferenceFrame
+
+__all__ = ['Joint', 'PinJoint']
+
+
+class Joint(ABC):
+    """Abstract base class for all specific joints.
+
+    Explanation
+    ===========
+
+    A joint subtracts degrees of freedom from a body. This is the base class
+    for all specific joints and holds all common methods acting as an interface
+    for all joints. Custom joint can be created by inheriting Joint class and
+    defining all abstract functions.
+
+    The abstract methods are:
+
+    - ``_generate_coordinates``
+    - ``_generate_speeds``
+    - ``_orient_frames``
+    - ``_set_angular_velocity``
+    - ``_set_linar_velocity``
+
+    Parameters
+    ==========
+
+    name : string
+        A unique name for the joint.
+    parent : Body
+        The parent body of joint.
+    child : Body
+        The child body of joint.
+    coordinates: List of dynamicsymbols, optional
+        Generalized coordinates of the joint.
+    speeds : List of dynamicsymbols, optional
+        Generalized speeds of joint.
+    parent_joint_pos : Vector, optional
+        Vector from the parent body's mass center to the point where the parent
+        and child are connected. The default value is the zero vector.
+    child_joint_pos : Vector, optional
+        Vector from the parent body's mass center to the point where the parent
+        and child are connected. The default value is the zero vector.
+    parent_axis : Vector, optional
+        Axis fixed in the parent body which aligns with an axis fixed in the
+        child body. The default is x axis in parent's reference frame.
+    child_axis : Vector, optional
+        Axis fixed in the child body which aligns with an axis fixed in the
+        parent body. The default is x axis in child's reference frame.
+
+    Attributes
+    ==========
+
+    name : string
+        The joint's name.
+    parent : Body
+        The joint's parent body.
+    child : Body
+        The joint's child body.
+    coordinates : list
+        List of the joint's generalized coordinates.
+    speeds : list
+        List of the joint's generalized speeds.
+    parent_joint : Point
```

Comment:
```suggestion
    parent_point : Point
```

Reply:
Doesn't `parent_joint` seems better?

Reply:
It returns a point, not a joint, so the name "point" seems to say what it is.

Reply:
But it represents the joint with parent body, that's why I'm inclined towards `parent_joint` , it's description also supports it.

Reply:
It is a good rule of thumb to name things what they are. Our definition of a "joint" is much than a "point". In the case of the pin joint a joint = a name, two points, and two axes! If a user types `my_joint.parent_joint` interactively they'd expect to get a joint object because of its name. There are only two points associated with a Joint object, one for the child and one for the parent. That's unambiguous.

Reply:
Okay, I'll update that.

Reply:
Done